### PR TITLE
Port .NET Native type name parser

### PR DESCRIPTION
**Pull request by mateoatr**
See #1387.

This adds the type name parsing logic used by CoreRT in reflection scenarios. With this we can successfully parse generic types, fixing scenarios such as #1469.

For reviewing I'd suggest paying particular attention to `ReflectionTypeNameResolver` in TypeParsing.cs; all other new files were ported from CoreRT with insignificant modifications.

### CODE REVIEW

File: src/linker/Linker/TypeParsing.cs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+					var foundType = ResolveTypeName (assemblyDefinition, assemblyQualifiedTypeName.TypeName);
+					if (foundType == null)
+						continue;
+
+					return foundType;
+				}
+
+				return null;
+			} else if (typeName is NonQualifiedTypeName) {
+				return ResolveTypeName (null, (NonQualifiedTypeName) typeName);
```
Comment (vitek-karas) [inline src/linker/Linker/TypeParsing.cs:52]: This explicitly calls ResolveTypeName with assembly null, but the method below will NullRef on null assembly.

File: src/linker/Linker.Steps/MarkStep.cs [hunk 2]
```diff
@@ -1548,25 +1547,15 @@ protected virtual void DoAdditionalInstantiatedTypeProcessing (TypeDefinition ty
...
-					targetTypeReference = (TypeReference) property.Argument.Value;
-					break;
-				}
+				if (property.Name == "Target")
+					return ((TypeReference) property.Argument.Value)?.Resolve ();
 
-				if (property.Name == "TargetTypeName") {
-					if (TypeNameParser.TryParseTypeAssemblyQualifiedName ((string) property.Argument.Value, out string typeName, out string assemblyName)) {
-						if (string.IsNullOrEmpty (assemblyName))
-							targetTypeReference = asm.MainModule.GetType (typeName);
```
Comment (vitek-karas) [inline src/linker/Linker.Steps/MarkStep.cs:1561]: If the type name is not assembly qualified, this will resolve it against the current assembly.
But the new code will search all assemblies and return the first hit. For things like Nullable attributes which are added into each assembly - the behavior will suddenly be very different.

File: test/Mono.Linker.Tests/Tests/TypeNameParserTests.cs [hunk 3]
```diff
@@ -1,125 +0,0 @@
-﻿using NUnit.Framework;
-
-namespace Mono.Linker.Tests
-{
-	[TestFixture]
-	public class TypeNameParserTests
```
Comment (vitek-karas) [inline test/Mono.Linker.Tests/Tests/TypeNameParserTests.cs:6]: We should try to adapt these test cases - even if the code is very different the tests themselves seem to be valid.

**Review by vitek-karas — CHANGES_REQUESTED** [review]
We should try hard to avoid changing existing behavior. I didn't dig deep enough to tell how much of a difference there is between the old and new algorithms, but at least the visible differences around assembly resolution should be removed.
Also - we should take this opportunity to add more tests - the fact that none of the assembly differences caused test failures means that we're missing coverage.
Also - given the bug which started all this - we should have tests for that bug and similar cases as well.

File: src/linker/Linker/AssemblyNameHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+				throw new FileLoadException ();
+
+			StringBuilder sb = new StringBuilder ();
+			if (a.Name != null) {
+				sb.AppendQuoted (a.Name);
+			}
+
+			if (a.Version != null) {
+				Version canonicalizedVersion = a.Version.CanonicalizeVersion ();
+				if (canonicalizedVersion.Major != ushort.MaxValue) {
```
Comment (MichalStrehovsky) [inline src/linker/Linker/AssemblyNameHelpers.cs:29]: Did you make any changes to these files that we should review?

@marek-safar The fact that this repo uses formatting convention not used by the .NET teams makes source sharing frustrating.

File: src/linker/Linker/TypeParsing.cs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+			// This is unreachable
+			throw new NotImplementedException ();
+		}
+
+		public TypeDefinition ResolveTypeName (AssemblyDefinition assembly, NonQualifiedTypeName typeName)
+		{
+			if (typeName is ConstructedGenericTypeName) {
+				ConstructedGenericTypeName genericTypeName = (ConstructedGenericTypeName) typeName;
+				return assembly.MainModule.GetType (genericTypeName.GenericType.ToString ());
```
Comment (MichalStrehovsky) [inline src/linker/Linker/TypeParsing.cs:63]: Will this lose the generic arguments?

One of the things #1387 calls out is  `Type.GetType("Foo[Bar]"))` should make sure both Foo and Bar are kept. It doesn't look like that would happen here.

File: src/linker/Linker/AssemblyNameHelpers.cs [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// Copyright (c) Microsoft Corporation. All rights reserved.
+
+using System.IO;
+using System.Text;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace System.Reflection.Runtime.Assemblies
```
Comment (marek-safar) [inline src/linker/Linker/AssemblyNameHelpers.cs:11]: The namespace does not match the file path (and tbh it has quite odd name)

File: src/linker/Linker/TypeParsing.cs [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			return ResolveTypeName (parsedTypeName);
+		}
+
+		private TypeDefinition ResolveTypeName (TypeName typeName)
+		{
+			if (typeName is AssemblyQualifiedTypeName) {
+				AssemblyQualifiedTypeName assemblyQualifiedTypeName = (AssemblyQualifiedTypeName) typeName;
+				RuntimeAssemblyName assemblyName = assemblyQualifiedTypeName.AssemblyName;
+				foreach (var assemblyDefinition in _context.GetAssemblies ()) {
+					if (assemblyName != null && assemblyDefinition.Name.Name != assemblyName.Name)
```
Comment (marek-safar) [inline src/linker/Linker/TypeParsing.cs:40]: Why is `if (assemblyName != null` inside the loop?

File: src/linker/Linker/TypeParsing.cs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+			return ResolveTypeName (parsedTypeName);
+		}
+
+		private TypeDefinition ResolveTypeName (TypeName typeName)
+		{
+			if (typeName is AssemblyQualifiedTypeName) {
+				AssemblyQualifiedTypeName assemblyQualifiedTypeName = (AssemblyQualifiedTypeName) typeName;
+				RuntimeAssemblyName assemblyName = assemblyQualifiedTypeName.AssemblyName;
+				foreach (var assemblyDefinition in _context.GetAssemblies ()) {
```
Comment (marek-safar) [inline src/linker/Linker/TypeParsing.cs:39]: This should use `_context.GetLoadedAssembly (assemblyName.Name)`

File: src/linker/Linker/TypeParsing.cs [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+					if (assemblyName != null && assemblyDefinition.Name.Name != assemblyName.Name)
+						continue;
+
+					var foundType = ResolveTypeName (assemblyDefinition, assemblyQualifiedTypeName.TypeName);
+					if (foundType == null)
+						continue;
```
Comment (marek-safar) [inline src/linker/Linker/TypeParsing.cs:45]: I don't think `continue` will ever succeed

File: src/linker/Linker/TypeParsing.cs [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+			} catch (System.IO.FileLoadException) {
+				return null;
+			}
+
+			return ResolveTypeName (parsedTypeName);
+		}
+
+		private TypeDefinition ResolveTypeName (TypeName typeName)
+		{
+			if (typeName is AssemblyQualifiedTypeName) {
```
Comment (marek-safar) [inline src/linker/Linker/TypeParsing.cs:36]:
```suggestion
			if (typeName is AssemblyQualifiedTypeName typeName) {
```

**Review by marek-safar — CHANGES_REQUESTED** [review]

File: src/linker/Linker/AssemblyNameHelpers.cs (hunk 4 above)
Comment (marek-safar) [reply src/linker/Linker/AssemblyNameHelpers.cs:29]: I agree and this code should not land under `src/linker/` but under `external/<something>` where it can use any formatting convention and it's clear where it's coming from. I don't know where this code was copied from but it should also copy tests as I'm not comfortable landing hundreds of lines of a parser code without test coverage.

Comment (MichalStrehovsky) [reply src/linker/Linker/AssemblyNameHelpers.cs:29]: This is originally extracted from the runtime reflection stack. The tests for that are the Type.GetType tests in the runtime repo - converting them to something workable is probably a non-trivial amount of work. If we need to service this, the fix should go through the runtime.

Given this code shipped in several major versions of a .NET runtime and the .NET Native compiler, I have pretty high confidence in it even with light (integration-level) test coverage here.

File: src/linker/Linker/AssemblyNameHelpers.cs (hunk 6 above)
Comment (mateoatr) [reply src/linker/Linker/AssemblyNameHelpers.cs:11]: Should I change all the places where this namespace is used to `Mono.Linker`?

File: src/linker/Linker/AssemblyNameHelpers.cs (hunk 4 above)
Comment (mateoatr) [reply src/linker/Linker/AssemblyNameHelpers.cs:29]: @MichalStrehovsky 
 > Did you make any changes to these files that we should review?

Not really. Probably the most interesting change is this - how we parse the assembly version, which is a bit different from how we originally did it for .NET Native, however the code here is how we currently do it in CoreRT: https://github.com/dotnet/corert/blob/9d8bd29a71f68941aaa2c00e9c9a0eed86cafa10/src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs#L26-L52
That, and the fact that the `ToString` of `NestedTypeName` returns `DeclaringType + "/" Name` instead of `DeclaringType + "+" Name`, because of how Cecil encodes things.

File: test/Mono.Linker.Tests/Tests/TypeNameParserTests.cs (hunk 3 above)
Comment (mateoatr) [reply test/Mono.Linker.Tests/Tests/TypeNameParserTests.cs:6]: I adapted these.

File: src/linker/Linker.Steps/MarkStep.cs (hunk 2 above)
Comment (mateoatr) [reply src/linker/Linker.Steps/MarkStep.cs:1561]: @vitek-karas I made changes to this so that we honor the old (current) behavior.

@MichalStrehovsky I can change this back if it's too risky. I felt comfortable changing this since it seems that any type name that returns true when passed to `TryParseTypeAssemblyQualifiedName` should be correctly parsed by the new type name parser.

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 11]
```diff
@@ -1335,7 +1335,8 @@ void ProcessCreateInstanceByName (ref ReflectionPatternContext reflectionContext
 								continue;
 							}
 
-							var resolvedType = resolvedAssembly.FindType (typeNameStringValue.Contents);
+							var typeName = System.Reflection.Runtime.TypeParsing.TypeParser.ParseAssemblyQualifiedTypeName (typeNameStringValue.Contents).TypeName;
```
[the remark this answers is not in the document]
Comment (mateoatr) [reply src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs]: I did not. I changed this to be able to remove `FindType`.

File: src/linker/Linker/AssemblyNameHelpers.cs (hunk 6 above)
Comment (MichalStrehovsky) [reply src/linker/Linker/AssemblyNameHelpers.cs:11]: No - move the files that come from the runtime to the external directory like Marek suggests below (something like external\TypeParsing\System\Reflection\Runtime\Assemblies) and add a README.md to the root of TypeParsing that tells the source and commit ID that was used as the base. That way we'll have future reference.

Also undo the formatting changes on the files that come from outside so that this is easier to diff, should we have a need.

File: src/linker/Linker/TypeNameResolver.cs [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+			if (assembly == null)
+				return null;
+
+			if (typeName is AssemblyQualifiedTypeName assemblyQualifiedTypeName) {
+				return ResolveTypeName (assembly, assemblyQualifiedTypeName.TypeName);
+			} else if (typeName is ConstructedGenericTypeName constructedGenericTypeName) {
+				var genericTypeDef = ResolveTypeName (assembly, constructedGenericTypeName.GenericType)?.Resolve ();
+				var genericInstanceType = new GenericInstanceType (genericTypeDef);
+				foreach (var arg in constructedGenericTypeName.GenericArguments) {
+					genericInstanceType.GenericArguments.Add (ResolveTypeName (assembly, arg));
```
Comment (MichalStrehovsky) [inline src/linker/Linker/TypeNameResolver.cs:60]: If ResolveTypeName returns null, this will create a generic instance over null. That will probably crash things.

(Also for HasElementTypeName below.)

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 13]
```diff
@@ -797,13 +797,14 @@ public override bool HandleCall (MethodBody callingMethodBody, MethodReference c
...
 							if (typeNameValue is KnownStringValue knownStringValue) {
-								TypeDefinition foundType = AssemblyUtilities.ResolveFullyQualifiedTypeName (_context, knownStringValue.Contents);
+								TypeReference foundType = _context.TypeNameResolver.ResolveTypeName (knownStringValue.Contents);
 								if (foundType == null) {
 									// Intentionally ignore - it's not wrong for code to call Type.GetType on non-existing name, the code might expect null/exception back.
 									reflectionContext.RecordHandledPattern ();
 								} else {
-									reflectionContext.RecordRecognizedPattern (foundType, () => _markStep.MarkTypeVisibleToReflection (foundType, new DependencyInfo (DependencyKind.AccessedViaReflection, callingMethodDefinition), callingMethodDefinition));
-									methodReturnValue = MergePointValue.MergeValues (methodReturnValue, new SystemTypeValue (foundType));
+									var typeDef = foundType?.Resolve ();
```
Comment (vitek-karas) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:805]: This should be moved above the if and we should check that Resolve didn't return null. It's "easy" to create a type ref in Cecil - but there's no guarantee that it will actually resolve. If the Resolve fails here, we will likely nullref below.

File: src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs [hunk 14]
```diff
@@ -1344,7 +1345,7 @@ void ProcessCreateInstanceByName (ref ReflectionPatternContext reflectionContext
 								continue;
 							}
 
-							MarkConstructorsOnType (ref reflectionContext, resolvedType,
+							MarkConstructorsOnType (ref reflectionContext, resolvedType?.Resolve (),
```
Comment (vitek-karas) [inline src/linker/Linker.Dataflow/ReflectionMethodBodyScanner.cs:1348]: Same here - Resolve may return null - which needs to be handled.

File: external/TypeParsing/README.md [hunk 15]
```diff
@@ -0,0 +1 @@
+The code in this folder was adapted from dotnet/corert commit c8bfca5f4554badfb89b80d2319769f83512bf62
```
Comment (marek-safar) [inline external/TypeParsing/README.md:1]: The folder name should start with `external/corert/`

File: external/TypeParsing/System/Reflection/AssemblyNameFormatter.cs [hunk 16]
```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (marek-safar) [inline external/TypeParsing/System/Reflection/AssemblyNameFormatter.cs:1]: It might be better to match the file path as well (e.g. `src/System.Private.CoreLib/shared/System/Reflection/AssemblyNameFormatter.cs` in this case)

File: src/linker/Linker/ModuleDefinitionExtensions.cs [hunk 17]
```diff
@@ -0,0 +1,14 @@
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	public static class ModuleDefinitionExtensions
+	{
+
+		public static bool IsCrossgened (this ModuleDefinition module)
```
Comment (marek-safar) [inline src/linker/Linker/ModuleDefinitionExtensions.cs:8]: Why is it needed?

Comment (mateoatr) [reply src/linker/Linker/ModuleDefinitionExtensions.cs:8]: This used to be in `AssemblyResolver`, I simply splitted that file to fit `<Type>Extensions.cs`. Looks like this is used for writing back pure IL for crossgen'd assemblies: https://github.com/mono/linker/blob/master/src/linker/Linker.Steps/OutputStep.cs#L117-L118

**Review by vitek-karas — APPROVED** [review]

**Review by marek-safar — APPROVED** [review]
